src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -105,26 +110,51 @@ const std::shared_ptr<SQLQueryPlan>& SQLPipelineStatement::get_query_plan() {
     return _query_plan;
   }
 
-  const auto& lqp = get_optimized_logical_plan();
+  // If we need a transaction context but haven't passed one in, this is the latest point where we can create it
+  if (!_transaction_context && _use_mvcc) {
+    _transaction_context = TransactionManager::get().new_transaction_context();
+  }
+
   _query_plan = std::make_shared<SQLQueryPlan>();
 
-  const auto started = std::chrono::high_resolution_clock::now();
+  auto started = std::chrono::high_resolution_clock::now();
+
+  // Handle query plan if statement has been cached
+  const auto cached_plan = SQLQueryCache<SQLQueryPlan>::get().try_get(_sql_string);
+  if (cached_plan) {
```

Comment:
Personally, I like `if (const auto cached_plan = SQLQueryCache<SQLQueryPlan>::get().try_get(_sql_string);)` so that the scope is kept small, but that's just a suggestion.